### PR TITLE
`eth_getLogs` does not validate the topics length

### PR DESCRIPTION
**Pull request by m-Peter**
Closes: https://github.com/onflow/flow-evm-gateway/issues/422

## Description

In addition to the `topics` length validation, we also add more tests regarding the matching logic.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels 

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling in log retrieval to improve reliability.
	- Introduced validation checks for the maximum number of allowed topics when creating filters.

- **Bug Fixes**
	- Improved test cases for log filtering to cover additional scenarios and error conditions.

- **Tests**
	- Refactored and expanded test cases for log filtering mechanisms to enhance robustness and clarity.

### CODE REVIEW

File: services/logs/filter_test.go [hunk 1]
```diff
@@ -168,33 +168,71 @@ func TestIDFilter(t *testing.T) {
 		desc: "wildcard all logs",
 		id:   mustHash(blocks[0]),
 		criteria: FilterCriteria{
-			Addresses: []common.Address{lgs[0].Address},
```
Comment (sideninja) [inline services/logs/filter_test.go:171]: why is filter by address removed? we need to have one test at least that filters by address

File: services/logs/filter_test.go [hunk 2]
```diff
@@ -168,33 +168,71 @@ func TestIDFilter(t *testing.T) {
...
+			Addresses: []common.Address{},
 			Topics:    [][]common.Hash{},
 		},
-		expectLogs: lgs[:2], // first two are all logs from the address
+		expectLogs: logs[:], // block 0 has 3 logs in total
+	}, {
+		desc: "single topic no address match single log",
```
Comment (sideninja) [inline services/logs/filter_test.go:176]: nice

File: services/logs/filter.go [hunk 3]
```diff
@@ -124,13 +128,17 @@ func NewIDFilter(
 	criteria FilterCriteria,
 	blocks storage.BlockIndexer,
 	receipts storage.ReceiptIndexer,
-) *IDFilter {
+) (*IDFilter, error) {
+	if len(criteria.Topics) > maxTopics {
```
Comment (sideninja) [inline services/logs/filter.go:132]: good idea

**Review by sideninja — APPROVED** [review]
left a comment on the removed test

File: services/logs/filter_test.go (hunk 1 above)
Comment (m-Peter) [reply services/logs/filter_test.go:171]: We already have such a test below, see: https://github.com/onflow/flow-evm-gateway/pull/423/files#diff-3dcf94bdc0a09d411408b6d264b30c0f2ad933cdf8dcf518d30a6718640a97a6R222-R227 .
It's just that the `wildcard all logs` filter was basically a duplicate test, as it provided an address.
So it wasn't really a wildcard filter criteria.

**Comment by sideninja** [comment]
Very nice PR, will merge now.

**Pull request merged**